### PR TITLE
Virology QoL update (+ some refactoring)

### PR DESCRIPTION
**Pull request by Absolucy**
## About The Pull Request

This PR does many things related to virology, mostly the viral extrapolator, but also some other virology-related things (and one thing not very related to virology I decided to mash in because I was touching that code anyways)

- Made the viral extrapolator output look much better.
- Added some additional info to the extrapolator's examine - telling the user the maximum stealth / level it can extract, and how long extractions take.
- The viral extrapolator will now tell you when scanning if it has extracted that same virus previously.
- You no longer have to click on an exact sprite to scan or extract a virus from something - if you click on something that doesn't have any valid diseases (such as a floor tile), you will instead scan/extract whatever thing on that tile that _does_ have a valid disease. (And if there's multiple things on the tile, you can choose which one to scan from a popup)
- Extracting a virus from someone now displays a visible message, rather than being completely silent except for the beep afterwards.
- Vector (the virologist's pet hamster) is always guaranteed to have at least one level 9 symptom in its freebie virus.
- Morgue corpses and _roundstart_ monkeys can also generate with dormant diseases now, although they can be less reliable and more of a wildcard than diseases in dried blood/vomit/gibs or crew members.
- Fixed only being able to extract symptoms rather than full diseases from rats.
- Morgue corpses now have a small chance to be a non-human (and non-IPC) roundstart race.
- Refactored how `extrapolator_act` works.

## Why It's Good For The Game

QoL good, plus gives virologists more chances to get good rolls in a heavily RNG-based job.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

23-07-07-1688703949-dreamseeker
23-07-07-1688703891-dreamseeker

23-07-07-1688703851-dreamseeker
23-07-07-1688703788-dreamseeker

23-07-07-1688703333-dreamseeker
23-07-07-1688703206-dreamseeker
23-07-07-1688703176-dreamseeker
23-07-07-1688703073-dreamseeker
23-07-07-1688703022-dreamseeker
23-07-07-1688703005-dreamseeker

</details>

## Changelog
:cl:
add: Made the viral extrapolator output look much better.
add: Added some additional info to the viral extrapolator's examine - telling the user the maximum stealth / level it can extract, and how long extractions take.
add: The viral extrapolator will now tell you when scanning if it has extracted that same virus previously.
tweak: You no longer have to click on an exact sprite to scan or extract a virus from something with the viral extrapolator - if you click on something that doesn't have any valid diseases (such as a floor tile), you will instead scan/extract whatever thing on that tile that does have a valid disease. (And if there's multiple things on the tile, you can choose which one to scan from a popup)
tweak: Extracting a virus from someone now displays a visible message, rather than being completely silent except for the beep afterwards.
add: Vector (the virologist's pet hamster) is always guaranteed to have at least one level 9 symptom in its freebie virus.
add: Morgue corpses and roundstart monkeys can also generate with dormant diseases now, although they can be less reliable and more of a wildcard than diseases in dried blood/vomit/gibs or crew members.
fix: Fixed only being able to extract symptoms rather than full diseases from rats.
fix: Morgue corpses now have a small chance to be a non-human (and non-IPC) roundstart race.
refactor: Refactored how `extrapolator_act` works.
/:cl:

### CODE REVIEW

**Comment by CydiaLamiales** [comment]
While you're at it, could you actually make Vector always have a lvl 9 symptom?

**Comment by Absolucy** [comment]
@CydiaLamiales done!

File: code/game/objects/items/devices/scanners.dm [hunk 1]
```diff
@@ -1031,99 +1081,221 @@ GENE SCANNER
...
+			return COLOR_MAROON
 		else
-			to_chat(user, "<span class='info'><font color='green'><b>[D.name]</b>, stage [D.stage]/[D.max_stages].</font></span>")
-
-/obj/item/extrapolator/proc/extrapolate(atom/AM, var/list/diseases = list(), mob/user, isolate = FALSE, timer = 100)
-	var/list/advancediseases = list()
-	var/list/symptoms = list()
-	if(using)
-		to_chat(user, "<span class='warning'>The extrapolator is already in use.</span>")
+			return COLOR_SILVER
```
Comment (EvilDragonfiend) [inline code/game/objects/items/devices/scanners.dm]: avoid using direct color defines. You should use class name css colors, and add acceptable color to dark/light chat themes.

**Review by EvilDragonfiend — CHANGES_REQUESTED** [review]

File: code/game/objects/items/devices/scanners.dm (hunk 1 above)
Comment (Absolucy) [reply code/game/objects/items/devices/scanners.dm]: ... Do you know any spans which would fit well here?

Comment (Absolucy) [reply code/game/objects/items/devices/scanners.dm]: Meh, I just removed the colors.

File: code/modules/jobs/job_types/_job.dm [hunk 2]
```diff
@@ -220,7 +220,7 @@
 
 	if(!visualsOnly && announce)
 		announce(H)
-	dormant_disease_check(H)
+	H.give_random_dormant_disease(biohazard, min_level = CHECK_BITFIELD(flag, CLOWN | MIME) ? 0 : 4)
```
Comment (EvilDragonfiend) [inline code/modules/jobs/job_types/_job.dm]:
```suggestion
	if(title == JOB_NAME_CLOWN || title == JOB_NAME_MIME)
		H.give_random_dormant_disease(biohazard, min_level = 0)
	else
		H.give_random_dormant_disease(biohazard, min_level = 4)
```
Do not use this bitflag. That should be removed.

File: code/modules/mob/living/simple_animal/friendly/mouse.dm [hunk 3]
```diff
@@ -27,10 +27,12 @@
...
 	var/chew_probability = 1
 	chat_color = "#82AF84"
-	var/list/ratdisease = list()
 	can_be_held = TRUE
 	worn_slot_flags = ITEM_SLOT_HEAD
-
+	/**
+	 * A list of diseases carried by this rat.
+	 */
+	var/list/datum/disease/rat_diseases = list()
```
Comment (EvilDragonfiend) [inline code/modules/mob/living/simple_animal/friendly/mouse.dm:33]: /// will be sufficient for the comment

File: code/game/objects/items/devices/scanners.dm [hunk 4]
```diff
@@ -985,36 +985,86 @@ GENE SCANNER
...
+	 * The extrapolator can extract any symptom with a stealth below this value.
+	 */
+	var/maximum_level = 7
+	/**
+	 * The typepath of the default scanning module that will generate in the extrapolator, if it starts with none.
+	 */
+	var/default_scanning_module = /obj/item/stock_parts/scanning_module
+	/**
+	 * Cooldown for when the extrapolator can be used next.
+	 */
```
Comment (EvilDragonfiend) [inline code/game/objects/items/devices/scanners.dm]: please use /// these are too verbose

**Review by EvilDragonfiend — CHANGES_REQUESTED** [review]
I need to understand how this code works more but I can't invest more time more, but here are some change requests.

**Comment by EvilDragonfiend** [comment]
`EXTRAPOLATOR_SPECIAL_ISOLATE` seems to be checked by nowhere. Why should a return have it?

**Comment by Absolucy** [comment]
@EvilDragonfiend https://github.com/BeeStation/BeeStation-Hornet/blob/b9893651b1dcdc7ec08df0fe6ca2d386432829e9/code/game/objects/items/devices/scanners.dm#L1168-L1169

**Review by EvilDragonfiend — CHANGES_REQUESTED** [review]
I don't like the return format, and everything is mixed in disease list. I really don't feel right that.
This is my suggestion to change return
```DM
#define EXTRAPOLATOR_RESULT_DISEASES extrapolator_result_disease
#define EXTRAPOLATOR_RESULT_ACT_PRIORITY extrapolator_result_action_priority
#define EXTRAPOLATOR_ACT_PRIORITY_SPECIAL extrapolator_action_priority_special
#define EXTRAPOLATOR_ACT_PRIORITY_ISOLATE extrapolator_action_priority_isolate
#define EXTRAPOLATOR_ACT_CHECK(target_list, wanted_action_priority) (##target_list[EXTRAPOLATOR_RESULT_ACT_PRIORITY] == ##wanted_action_priority)

/atom/proc/extrapolator_act()
	. = list(EXTRAPOLATOR_RESULT_DISEASES = list())

/obj/item/reagent_containers/extrapolator_act(mob/living/user, obj/item/extrapolator/extrapolator, dry_run = FALSE)
	// Always attempt to isolate diseases from reagent containers, if possible.
	. = ..()
	.[EXTRAPOLATOR_RESULT_ACT_PRIORITY] = EXTRAPOLATOR_ACT_PRIORITY_ISOLATE
	var/datum/reagent/blood/blood = locate() in reagents.reagent_list
	if(blood)
		.[EXTRAPOLATOR_RESULT_DISEASES] |= blood.get_diseases()

/mob/living/simple_animal/hostile/blob/blobspore/extrapolator_act(mob/living/user, obj/item/extrapolator/extrapolator, dry_run = FALSE)
	. = ..() | spore_diseases
	if(!dry_run && EXTRAPOLATOR_ACT_CHECK(., EXTRAPOLATOR_ACT_PRIORITY_SPECIAL) && extrapolator.create_culture(user, spore_diseases))
		user.visible_message("<span class='danger'>[user] stabs [src] with [extrapolator], sucking it up!</span>", \
				"<span class='danger'>You stab [src] with [extrapolator]'s probe, destroying it!</span>")
		dust()
		.[EXTRAPOLATOR_RESULT_ACT_PRIORITY] = EXTRAPOLATOR_ACT_PRIORITY_SPECIAL

/obj/item/extrapolator/afterattack(atom/target, mob/living/user, proximity_flag, click_parameters)
	. = ..()
	if(!proximity_flag && !scan)
		return
	if(using)
		to_chat(user, "<span class='warning'>[icon2html(src, user)] The extrapolator is already in use.</span>")
		return
	if(!COOLDOWN_FINISHED(src, usage_cooldown))
		to_chat(user, "<span class='warning'>[icon2html(src, user)] The extrapolator is still recharging!</span>")
		return
	if(scanner)
		var/list/extrapolator_result = target?.extrapolator_act(user, src, dry_run = TRUE)
		if(!extrapolator_result) // it even didn't return as a list
			return
		if(!length(extrapolator_result[EXTRAPOLATOR_RESULT_DISEASES]))
			var/list/atom/targets = find_valid_targets(user, target)
			var/target_amt = length(targets)
			if(target_amt)
				target = target_amt > 1 ? tgui_input_list(user, "Select object to analyze", "Viral Extrapolation", targets, default = targets[1]) : targets[1]
			if(target)
				diseases = target.extrapolator_act(user, src, dry_run = TRUE)
		if(!target)
			return
		if(!length(diseases))
			if(scan)
				to_chat(user, "<span class='notice'>[icon2html(src, user)] \The [src] fails to return any data.</span>")
			else
				to_chat(user, "<span class='notice'>[icon2html(src, user)] \The [src]'s probe detects no diseases.</span>")
			return
		if(EXTRAPOLATOR_ACT_CHECK(extrapolator_result, EXTRAPOLATOR_ACT_PRIORITY_SPECIAL))
			// extrapolator_act did some sort of special behavior, we don't need to do anything further
			return
		if(scan)
			scan(user, target)
		else
			extrapolate(user, target)
	else
		to_chat(user, "<span class='warning'>The extrapolator has no scanner installed!</span>")

```

so, it's a list in a list
`return_list[EXTRAPOLATOR_RESULT_DISEASES]` will return disease list
`return_list[EXTRAPOLATOR_RESULT_ACT_PRIORITY]` will tell which action this extrapolator should do.



EXTRAPOLATOR_ACT_CHECK define might not be necessary, but very long long define comparison looks bad lol. It's only for aesthetic reason.

**Review by PowerfulBacon — APPROVED** [review]

File: code/game/objects/items/devices/scanners.dm [hunk 5]
```diff
@@ -1045,99 +1074,191 @@ GENE SCANNER
...
+		return
+	if(!COOLDOWN_FINISHED(src, usage_cooldown))
+		to_chat(user, "<span class='warning'>[icon2html(src, user)] The extrapolator is still recharging!</span>")
+		return
 	if(scanner)
-		if(!target.extrapolator_act(user, src, scan))
-			if(locate(/datum/component/infective) in target.datum_components)
-				return //so the failure message does not show when we can actually extrapolate from a component
+		var/list/result = target?.extrapolator_act(user, src, dry_run = TRUE)
+		var/list/diseases = result && result[EXTRAPOLATOR_RESULT_DISEASES]
```
Comment (EvilDragonfiend) [inline code/game/objects/items/devices/scanners.dm:1108]: That'll get "TRUE" not `result[EXTRAPOLATOR_RESULT_DISEASES]`

File: code/game/objects/items/devices/scanners.dm [hunk 6]
```diff
@@ -1045,99 +1074,191 @@ GENE SCANNER
...
+		if(!length(diseases))
+			var/list/atom/targets = find_valid_targets(user, target)
+			var/target_amt = length(targets)
+			if(target_amt)
+				target = target_amt > 1 ? tgui_input_list(user, "Select object to analyze", "Viral Extrapolation", targets, default = targets[1]) : targets[1]
+			if(target)
+				result = target.extrapolator_act(user, src, dry_run = TRUE)
+				diseases = result && result[EXTRAPOLATOR_RESULT_DISEASES]
```
Comment (EvilDragonfiend) [inline code/game/objects/items/devices/scanners.dm:1116]: here too

**Review by EvilDragonfiend — CHANGES_REQUESTED** [review]

**Comment by Absolucy** [comment]
@EvilDragonfiend 
image

**Review by EvilDragonfiend — APPROVED** [review]

**Pull request merged**